### PR TITLE
SUP-1051: Team Member resource Framework/Genqlient conversion

### PR DESCRIPTION
**Pull request by james2791**
## PR checklist:
- [x] `docs/` updated
- [x] tests added
- [x] an example added to `examples/` (useful to demo new field/resource)
- [x] `CHANGELOG.md` updated with pending release information

This PR also introduces a `node.graphql` - to its name, allows for the easy obtaining of any type from the GraphQL that implements the `Node` interface. The `getNode` query currently has a Team Member type handling inside (with the corresponding `TeamMemberValues` fragment in `team_member.graphql` (note this was the approach instead of aliasing each `node` subquery which might've re-generated the same code with different function signatures)

### CODE REVIEW

File: buildkite/resource_team_member.go [hunk 1]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
+	if req.ProviderData == nil {
+		return
+	}
+
+	tm.client = req.ProviderData.(*Client)
+}
+
+func (teamMemberResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = resource_schema.Schema{
+		MarkdownDescription: "A team member resourc allows for the management of team team membership for existing organization users.",
```
Comment (mcncl) [inline buildkite/resource_team_member.go:48]: resource* and double "team"

File: buildkite/resource_team_member.go [hunk 2]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 			},
-			"user_id": &schema.Schema{
-				Required: true,
-				Type:     schema.TypeString,
+			"team_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the team to add/remove the user to/from.",
```
Comment (mcncl) [inline buildkite/resource_team_member.go:64]: Is there a function to remove the user from a team? Or does simple removing it from the config remove them from the team? Not sure this description is clear.

File: buildkite/resource_team_member.go [hunk 3]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
 			},
-			"uuid": &schema.Schema{
-				Computed: true,
-				Type:     schema.TypeString,
+			"user_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the user to add/remove.",
```
Comment (mcncl) [inline buildkite/resource_team_member.go:68]: Same. Would simply not providing a `user_id` remove that user from the team? When setting this value, I don't think I set it if I want to `remove`.

File: buildkite/resource_team_member.go [hunk 4]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
-
-	var mutation struct {
-		TeamMemberCreate struct {
-			TeamMemberEdge struct {
-				Node TeamMemberNode
-			}
-		} `graphql:"teamMemberCreate(input: {teamID: $teamId, userID: $userId})"`
-	}
+func (tm *teamMemberResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan, state teamMemberResourceModel
```
Comment (mcncl) [inline buildkite/resource_team_member.go:82]: If it's just a Create, you don't need 2 iterations of the Model. Maybe just use `plan` for clarity as on Create there is no `state`.

File: buildkite/resource_team_member.go [hunk 5]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
-	return nil
+	// Update state with values from API response/plan
+	state.Id = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Id)
+	state.Uuid = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Uuid)
+	state.TeamId = plan.TeamId
+	state.UserId = plan.UserId
+	// The role of the user will by default be "MEMBER" if none was entered in the plan
+	state.Role = types.StringValue(string(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Role))
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
```
Comment (mcncl) [inline buildkite/resource_team_member.go:114]: `&plan` would be clearer here; you're setting state to the values of `&THING`.

File: buildkite/graphql/node.graphql [hunk 6]
```diff
@@ -0,0 +1,9 @@
+query get(
```
Comment (mcncl) [inline buildkite/graphql/node.graphql:1]: Just `get` as a query name doesn't provide clarity on what it's getting.

File: buildkite/resource_team_member.go [hunk 7]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
 
-	vars := map[string]interface{}{
-		"id": d.Id(),
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	err := client.graphql.Query(context.Background(), &query, vars)
+	log.Printf("Reading team member %s ...", state.Id.ValueString())
+	apiResponse, err := get(tm.client.genqlient, state.Id.ValueString())
```
Comment (mcncl) [inline buildkite/resource_team_member.go:127]: See comment in `graphql` file; consider a clearer function name.

File: buildkite/resource_team_member.go [hunk 8]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
-	vars := map[string]interface{}{
-		"id": graphql.ID(d.Id()),
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	err := client.graphql.Mutate(context.Background(), &mutation, vars)
+	log.Printf("Deleting team member with ID %s ...", id)
+	_, err := deleteTeamMember(tm.client.genqlient, id)
+
```
Comment (mcncl) [inline buildkite/resource_team_member.go:194]: Why not just `_, err := deleteTeamMember(tm.client.genqlient, state.ID.ValueString())`?

File: buildkite/resource_team_member_test.go [hunk 9]
```diff
@@ -22,9 +21,9 @@ func TestAccTeamMember_add_remove(t *testing.T) {
 				Config: testAccTeamMemberConfigBasic("MEMBER"),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					// Confirm the team member exists in the buildkite API
-					testAccCheckTeamMemberExists("buildkite_team_member.test", &resourceTeamMember),
+					testAccCheckTeamMemberExists("buildkite_team_member.test", &tm),
 					// Confirm the team member has the correct values in Buildkite's system
-					testAccCheckTeamMemberRemoteValues(&resourceTeamMember, "MEMBER"),
+					testAccCheckTeamMemberRemoteValues(&tm, "MEMBER"),
```
Comment (mcncl) [inline buildkite/resource_team_member_test.go:26]: Knit, but passing in `string` then `&reference` in the first function, but then `&reference` then `string` in the second is a little jarring.

File: examples/main.tf [hunk 10]
```diff
@@ -8,6 +8,12 @@ resource "buildkite_team" "test" {
   default_member_role = "MEMBER"
 }
 
+resource "buildkite_team_member" "member1" {
+  role = "MEMBER"
+	team_id = "VXfhnVUS78HavgtP55WhWGzT401guK38Vm9LMMeCgQD124m8xaKBRq0Fth=="
+	user_id = "VXNbwSA9hwVPpMgUXu1dWIDWf45ZwU6J7deETygiLUrKBg2TZBxuDr6aKj=="
+}
+
```
Comment (mcncl) [inline examples/main.tf:16]: So, this is fine, but I think going forward we could create both a team and a user and then dynamically add in the values rather than hardcode.

**Review by mcncl — CHANGES_REQUESTED** [review]
Some that could be changed, some that are future things.

File: buildkite/graphql/node.graphql (hunk 6 above)
Comment (jradtilbrook) [reply buildkite/graphql/node.graphql:1]: Yeah I like getNode

File: buildkite/graphql/team_member.graphql [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    user {
+        id
+    }
+    role 
+}
+
+mutation createTeamMember(
+    $teamID: ID!, 
+    $userID: ID!,
+    # @genqlient(omitempty: true)
```
Comment (jradtilbrook) [inline buildkite/graphql/team_member.graphql:16]: Whats the reasoning behind omitempty instead of pointer: true?

File: buildkite/resource_team_member.go (hunk 3 above)
Comment (jradtilbrook) [reply buildkite/resource_team_member.go:68]: Yeah removing this resource removes them from the team; it removes the link between user and team. So the descriptions here are simply to "add"

File: buildkite/resource_team_member.go [hunk 12]
```diff
@@ -3,174 +3,208 @@ package buildkite
...
 	}
 
-	return nil
+	// Update state with values from API response/plan
+	state.Id = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Id)
+	state.Uuid = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Uuid)
+	state.TeamId = plan.TeamId
+	state.UserId = plan.UserId
+	// The role of the user will by default be "MEMBER" if none was entered in the plan
+	state.Role = types.StringValue(string(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Role))
```
Comment (jradtilbrook) [inline buildkite/resource_team_member.go:112]: Is this still true? There used to be a bug in the API but I think its been resolved. If not specified, the default will be set to the default configured on the team. So this could possibly modify state and cause a subsequent apply to generate a diff which is something we don't want. We should add a test for that edge case. Happy to pair on that with you

**Review by jradtilbrook — CHANGES_REQUESTED** [review]

File: buildkite/graphql/node.graphql (hunk 6 above)
Comment (james2791) [reply buildkite/graphql/node.graphql:1]: Me thee actually, thanks for the callout - more to get around the generated Node type naming - but just a few characters longer anyway

File: buildkite/resource_team_member.go (hunk 3 above)
Comment (james2791) [reply buildkite/resource_team_member.go:68]: Adjusting these - thanks both

File: buildkite/resource_team_member.go (hunk 2 above)
Comment (james2791) [reply buildkite/resource_team_member.go:64]: Adjusting these - thanks both

File: buildkite/resource_team_member.go (hunk 1 above)
Comment (james2791) [reply buildkite/resource_team_member.go:48]: Strikes again - thanks: adjusting in to-come-commit

File: buildkite/resource_team_member.go (hunk 4 above)
Comment (james2791) [reply buildkite/resource_team_member.go:82]: Dropping to one model in the latest commit

File: buildkite/resource_team_member.go (hunk 5 above)
Comment (james2791) [reply buildkite/resource_team_member.go:114]: Dropping to one model in the latest commit

File: buildkite/resource_team_member.go (hunk 7 above)
Comment (james2791) [reply buildkite/resource_team_member.go:127]: Adjusting in latest commit

File: buildkite/resource_team_member.go (hunk 8 above)
Comment (james2791) [reply buildkite/resource_team_member.go:194]: Adjusting in latest commit

File: buildkite/resource_team_member_test.go (hunk 9 above)
Comment (james2791) [reply buildkite/resource_team_member_test.go:26]: Solid - flipping the `testAccCheckTeamMemberRemoteValues` params in the adjustment commit

File: examples/main.tf (hunk 10 above)
Comment (james2791) [reply examples/main.tf:16]: Fair - honestly these values are length 58 sequences for examples only in truth 👍

**Review by mcncl — DISMISSED** [review]

File: buildkite/resource_team_member.go (hunk 12 above)
Comment (james2791) [reply buildkite/resource_team_member.go:110]: Per the chat - I think leaving the resource as is from the older SDKv2 impl (with role always required) would keep this for a future improvement/update 
Happy for other thoughts though 👍

File: buildkite/graphql/team_member.graphql (hunk 11 above)
Comment (james2791) [reply buildkite/graphql/team_member.graphql]: More to get around the eventual case that we can omit it entirely (and then the API should create said TM with the default role for the team) - more below (and also an improvement case since the older resource had the role as required (not optional) 👍

**Review by jradtilbrook — APPROVED** [review]

**Comment by james2791** [comment]
@mcncl one more before I wrap this in 👍

**Review by mcncl — APPROVED** [review]

**Comment by james2791** [comment]
🚀
